src/analysis/inner_product_space/orientation.lean

```diff
@@ -74,11 +139,130 @@ begin
 end
 
 /-- `orientation.fin_orthonormal_basis` gives a basis with the required orientation. -/
-@[simp] lemma orientation.fin_orthonormal_basis_orientation {n : ℕ} (hn : 0 < n)
+@[simp] lemma fin_orthonormal_basis_orientation (hn : 0 < n)
   (h : finrank ℝ E = n) (x : orientation ℝ E (fin n)) :
   (x.fin_orthonormal_basis hn h).to_basis.orientation = x :=
 begin
   haveI := fin.pos_iff_nonempty.1 hn,
   haveI := finite_dimensional_of_finrank (h.symm ▸ hn : 0 < finrank ℝ E),
   exact ((std_orthonormal_basis _ _).reindex $ fin_congr h).orientation_adjust_to_orientation x
 end
+
+section volume_form
+variables [finite_dimensional ℝ E] [_i : fact (finrank ℝ E = n)] (o : orientation ℝ E (fin n))
+
+include _i o
+
+/-- The volume form on an oriented real inner product space, a nonvanishing top-dimensional
+alternating form uniquely defined by compatibility with the orientation and inner product structure.
+-/
+def volume_form : alternating_map ℝ E ℝ (fin n) :=
+begin
+  classical,
+  unfreezingI { cases n },
+  { haveI : subsingleton E := finite_dimensional.finrank_zero_iff.mp _i.out,
+    let opos : alternating_map ℝ E ℝ (fin 0) := alternating_map.const_of_is_empty ℝ E (1:ℝ),
+    let oneg : alternating_map ℝ E ℝ (fin 0) := alternating_map.const_of_is_empty ℝ E (-1:ℝ),
+    exact o.eq_or_eq_neg_of_subsingleton.by_cases (λ _, opos) (λ _, oneg) },
+  { exact (o.fin_orthonormal_basis n.succ_pos _i.out).to_basis.det }
+end
+
+omit _i o
+
+@[simp] lemma volume_form_zero_pos [_i : fact (finrank ℝ E = 0)] :
+  orientation.volume_form (positive_orientation : orientation ℝ E (fin 0))
+  = alternating_map.const_linear_equiv_of_is_empty 1 :=
+by simp [volume_form, or.by_cases, dif_pos]
+
+@[simp] lemma volume_form_zero_neg [_i : fact (finrank ℝ E = 0)] :
+  orientation.volume_form (-positive_orientation : orientation ℝ E (fin 0))
+  = alternating_map.const_linear_equiv_of_is_empty (-1) :=
+begin
+  dsimp [volume_form, or.by_cases],
+  apply if_neg,
+  have : fintype.card (fin 0) = finrank ℝ E := by simp [_i.out],
+  rw [← ne.def, (-positive_orientation).ne_iff_eq_neg positive_orientation this],
+end
+
+include _i o
+
+/-- The volume form on an oriented real inner product space can be evaluated as the determinant with
+respect to any orthonormal basis of the space compatible with the orientation. -/
+lemma volume_form_robust (b : orthonormal_basis (fin n) ℝ E) (hb : b.to_basis.orientation = o) :
+  o.volume_form = b.to_basis.det :=
+begin
+  unfreezingI { cases n },
+  { have : o = positive_orientation := hb.symm.trans b.to_basis.orientation_is_empty,
+    simp [volume_form, or.by_cases, dif_pos this] },
+  { dsimp [volume_form],
+    rw [same_orientation_iff_det_eq_det, hb],
+    exact o.fin_orthonormal_basis_orientation _ _ },
+end
+
+attribute [irreducible] orientation.volume_form
```

Comment:
If I understand correctly, we should avoid adding after the fact irreducible attributes, as this is not possible in Lean 4. Can you move that attribute to the definition (and fix the lemmas inbetween by rewriting the definition when needed)?